File: app/src/main/java/com/softaai/designpatterns/java/behavioral/ChainOfResposibility/AbstractLogger.java
package com.softaai.designpatterns.java.behavioral.ChainOfResposibility;

/**
 * Created by Amol Pawar on 05-04-2019.
 * softAai Apps
 */
public abstract class AbstractLogger {
    public static int INFO = 1;
    public static int DEBUGE = 2;
    public static int ERROR = 3;
    protected int level;

    //next element in chain at responsibility
    protected AbstractLogger nextLogger;
    public void setNextLogger(AbstractLogger nextLogger){
        this.nextLogger = nextLogger;
    }

    public void logMessage(int level, String message){
        if(this.level <= level){
            write(message);
        }
        if(nextLogger != null){
            nextLogger.logMessage(level, message);
        }
    }

    abstract protected void write(String message);
}

File: app/src/main/java/com/softaai/designpatterns/kotlin/behavioral/AbstractLogger.kt
package com.softaai.designpatterns.kotlin.behavioral


/**
 * Created by Amol Pawar on 05-04-2019.
 * softAai Apps
 */
abstract class AbstractLogger {

    companion object {
        val INFO : Int = 1
        val DEBUG : Int = 2
        val ERROR : Int = 3
    }

    protected var level : Int = 0

    protected var nextAbstractLogger : AbstractLogger? = null

    fun setNextLogger(nextLogger : AbstractLogger){
        this.nextAbstractLogger = nextLogger
    }

    fun logMessage(level : Int, message : String){
        if(this.level <= level){
            write(message)
        }
        if(nextAbstractLogger != null){
            nextAbstractLogger?.logMessage(level, message)
        }
    }


    abstract protected fun write(message : String)
}

